### DISCUSSION TITLE
Proposed Content Versioning Enhancement for Directus

### DISCUSSION THREAD

**Discussion by kl-yy** [directus/directus#20372] [Feature Requests]
### Summary

Revised Content Versioning Enhancement Request for Directus

Current Behavior

- Currently, all users can promote a version to main if the admin has enabled versioning for a collection.

Desired Behavior

- Require a permission for users to promote a version to main.
- Allow admins to set the main version as read-only for a user.

### Motivation

The proposed changes would provide the following benefits:

- Improved security and data integrity: Requiring a permission to promote a version to main would help to prevent accidental or unauthorized changes to the main version of content. This would improve the security and data integrity of Directus.

- Enhanced collaboration: Setting the main version as read-only for a user would allow admins to prevent users from making changes to the main version of content, while still allowing them to view and edit the content in draft versions. This would facilitate collaboration between users and help to ensure that changes to content are reviewed and approved before they are published.

### Detailed Design

Following changes could be made to Directus to implement the desired behavior:

- Add a new permission called content_version_promote_main.

- Update the Content Versions UI to display a warning message to users who try to promote a version to main without the content_version_promote_main permission.

- Add a new setting to the Content Versions UI that allows admins to set the main version as read-only for a user.

### Must Have:
Require a permission for users to promote a version to main.
Allow admins to set the main version as read-only for a user.

### Should Have:
- Update the Content Versions UI to display a warning message to users who try to promote a version to main without the content_version_promote_main permission.
- Add the ability for users to see a list of all versions of a content item, including previous versions that have been replaced.
- Add the ability for users to diff between different versions of a content item.

### Could Have:
- Add the ability for users to restore a previous version of a content item.
- Add the ability for users to create branches of content items.
- Add the ability for users to create workflows for promoting content to main.

### Won't Have:

- Integrate with external systems.

### Drawbacks

Drawbacks of implementing the proposed content versioning enhancement in Directus

- Implementation cost: The proposed enhancement would require some additional code to be written and integrated into Directus. This would increase the complexity of the codebase and could make it more difficult to maintain and update.

- User space implementation: It is not clear whether the proposed enhancement could be implemented entirely in user space. If it requires changes to the core Directus code, this could make it more difficult to maintain and update.

- Impact on teaching Directus: The proposed enhancement could add some complexity to the Directus learning curve. Users would need to learn about the new permissions and settings that are associated with content versioning.

- Integration with existing and planned features: The proposed enhancement would need to be integrated with other existing and planned features in Directus. This could be challenging, and it is important to ensure that the enhancement does not break any existing functionality.

- Migration cost for existing Directus applications: If the proposed enhancement is a breaking change, it could require existing Directus applications to be migrated to the new version of Directus. This could be a costly and time-consuming process.

Overall, the proposed content versioning enhancement would add some complexity to Directus and could impact the learning curve for new users. It is important to carefully consider the implementation cost, user space implementation, integration with existing and planned features, and migration cost before implementing this enhancement.

### Alternatives

Alternatives to the proposed content versioning enhancement in Directus

- Implement a third-party content versioning plugin: There are a number of third-party content versioning plugins available for Directus. These plugins can provide similar functionality to the proposed enhancement, but they may not be as tightly integrated with Directus as a native enhancement would be.

- Use a separate content management system (CMS) with built-in content versioning: There are a number of CMSs that offer built-in content versioning. These CMSs may provide a more comprehensive set of features for managing content versions than Directus would with the proposed enhancement.

- Use a version control system (VCS) to manage content versions: VCSs, such as Git or Mercurial, can be used to manage content versions. This can be a good option for organizations that already use VCSs for other purposes.

Impact of not doing this

If the proposed content versioning enhancement is not implemented, Directus users will not have the ability to require a permission to promote a version to main or to set the main version as read-only for a user. This could make it more difficult to ensure the security and integrity of content, and it could also make it more difficult to collaborate on content.

### Breaking Change

The proposed enhancement is a breaking change. It would require changes to the core Directus code, and it would also require changes to the way that existing Directus applications manage content versions.

### Migration

It is possible to write a migration script to help existing Directus applications migrate to the new content versioning system. However, this migration script would be complex, and it would require careful testing to ensure that it does not break any existing functionality.

### Impact on Other Projects in the Directus Ecosystem

The proposed enhancement would have a significant impact on other projects in the Directus ecosystem. Any third-party plugins or modules that interact with content versioning would need to be updated to support the new system. Additionally, any existing Directus applications that use third-party plugins or modules would need to be updated to use the new versions of those plugins or modules.

### Recommended Adoption Strategy

The following is a recommended adoption strategy for the proposed content versioning enhancement:

- Create a migration script. The first step is to create a migration script that can be used to update existing Directus applications to the new content versioning system. This migration script should be carefully tested to ensure that it does not break any existing functionality.

- Update third-party plugins and modules. Any third-party plugins or modules that interact with content versioning should be updated to support the new system.

- Update existing Directus applications. Any existing Directus applications that use third-party plugins or modules should be updated to use the new versions of those plugins or modules.

- Release a new version of Directus. Once the migration script, third-party plugins, and existing Directus applications have been updated, a new version of Directus can be released.

- Encourage developers to adopt the new content versioning system. Developers should be encouraged to adopt the new content versioning system when they create new Directus applications.

- Provide support for developers who are migrating to the new content versioning system. The Directus team should provide support for developers who are migrating to the new content versioning system. This support could be provided in the form of documentation, tutorials, or one-on-one assistance.

By following this recommended adoption strategy, the Directus team can help to minimize the impact of the proposed content versioning enhancement on existing Directus developers and projects.

**Comment by Syed-Sheharyar** [comment]
### It does make sense

**Comment by luismts** [comment]
### **Choose who can promote a version**

Having the ability to select who can create a draft version of a content or promote it base on the right permissions would improve the management of some scenario where you don't want everybody publishing of affecting your content before its approval. This is a great featured for the content you need to review before affecting the content already published.

cc @rijkvanzanten

**Reply by luismts** [reply]
### Enhancements and Desired Features for Directus
I have created a thread with a list of features and enhancements that could further drive the utility and versatility of Directus.

You can see it [here](https://gist.github.com/luismts/ed40fb3ce57d0697f604a8ef59a2c875).

**Reply by rijkvanzanten** [reply]
Great doc! Thanks 🚀 (cc @alexchopin)

**Comment by d3v1an7** [comment]
This sounds really great! When thinking about how I might implement this in a newsroom context, it'd be ideal if the permission could also be aware of state/status (or other meta) in the asset.

For example, in our newsroom, before an asset is published, it doesn't really matter who can/should edit `main` -- but _after_ the asset is in a published state, tighter permissions would be ideal.

I haven't implemented versioning yet, so if the above changes the scope too much, I guess a work around would be to make `main` always readonly for all users, and the act of publishing (controlled via a flow) is what has permission to merge into main?

**Comment by luismts** [comment]
Hi @rijkvanzanten,

Any updates on this feature? It has more than 15 likes as requested for feature requests.

**Comment by ChristianDC13** [comment]
Hello, this feature has already received a significant number of votes, and it's clear that the community is eager to see something like this integrated. If it's not too much trouble, I would like to volunteer to work on the implementation. It would be a pleasure for me to contribute.

**Reply by luismts** [reply]
@rijkvanzanten let us know how we can contribute to this.

**Reply by luismts** [reply]
@rijkvanzanten this thread had so much love. And we would like to contribute, but at least we want to know how we can make a proper PR and if we can discuss the feature before we start the development process.

**Comment by ryanleung96** [comment]
Hi all, I've been working on this feature as I have a project with similar needs. I tried to preserve the original UIs from content versioning so my users wouldn't have a steep learning curve. Here are some screenshots:

![image](https://github.com/user-attachments/assets/a5438651-a37d-4af9-b27d-f2b629238b29)
![image](https://github.com/user-attachments/assets/29902b91-d17f-49c0-8195-6c33265123a9)

**Reply by luismts** [reply]
Hi @ryanleung96 that's fantastic! Did you make a public plugin? I would love to contribute. 

Also, I would like more information about what you have there. Can you describe the flow (how it works step by step)?

cc @ChristianDC13

**Reply by ryanleung96** [reply]
@luismts Those changes was coded in the Directus app and api, not as a standalone plugin. It involves new permission to review/approve a content type, and a flag to enable/disable the approval feature.

I would love to make it as a public plugin but the project has discontinued. I can restart the development some time later.

**Reply by luismts** [reply]
@ryanleung96  Can you share the code? So, we can see if we can make a plugin with it or if it has all the requirements mention before.

**Reply by luismts** [reply]
I mean, this is an open source project, and its purpose is to contribute to the community.

**Comment by birdream** [comment]
hey guy, any update on this feature? It would be really helpful if we can have that function.

**Reply by ryanleung96** [reply]
@birdream
Previously I had a solution but it requires an extensive alteration to the directus core. It gets the job done but comes with different issues. 

I had very limited knowledge on directus plugin system so I can’t really work on a good solution with plugins approach. Right now my interim solution has no updates.

**Comment by MarkBurvs** [comment]
I've just been looking in to the content version feature myself too as we have a need to allow users to create and edit versions, but NOT to promote them to main or to be able to edit main itself directly, whereas only admins should have rights to edit main directly or promote versions. 

This feels like such a massive missing feature that I've just spent hours fiddling around with permissions convinced I must be missing something. 

The closest I can get is: 
- Removing Update permissions on the collection to users. 
- Granting Create and Update permissions on directus_versions and directus_revisions to users.

That allows users to be able to create a new version of an item without being able to promote it, but they can't actually edit the fields at all, which renders it useless.

It feels like we also really need a way to filter a collection view by items that have un-promoted versions. This would allow an admin to easily review all suggested changes.

**Reply by MarkBurvs** [reply]
After a bit more fiddling, I think I have found a configuration that works. It requires having a status field that you use to set permissions on whether a user can edit an item (and by extension, promote a version) or not.

**Requirements**
We have two users, AdminUser and GeneralUser. 
We have a Collection named "articles" which contains a Status field (Published or Draft).

We want AdminUser:
- To be able to create and publish new articles.
- To be able to review proposed new versions of existing articles and promote them if necessary.

We want GeneralUser:
- to be able to create new draft articles but not publish them.
- to be prevented from making updates to published articles. 
- to be able to propose updates to published articles by making new versions of them using the Content Versioning system..
- to be prevented from promoting a version to the "main" version of an article.

**Setup**
Enable Content Versioning on the Articles collection.
Make sure the Articles collections has a status field with options of "Published" and "Draft".

Ensure AdminUser has full CRUD permissions on the Articles collection.

But for our GeneralUser, setup their permissions on Articles as follows:
- Full Read access.
- Custom Create Access. Under Field Permissions, tick all fields. Under Field Validation, add a rule that status equals Draft.
- Custom Update Access. Under Item Permissions, create a rule that status Doesn’t equal Published. 
- Under Field Permissions, tick all fields apart from status.

Also for GeneralUser:
- Grant full CRUD permissions on the directus_versions collection.

That should do it!
It’s not ideal, and certainly not an obvious setup. But appears to do what I need it to do. 
It’s a shame that there’s no way of visually telling when an item has one or more alternate versions without going into the item and clicking on the word “Main” at the top, which also feels like an oversight in the design of the versioning system. But I reckon I can probably do something with Flows to alert the AdminUsers when a new version gets created.
